dipy/align/vector_fields.pyx

```diff
@@ -3099,71 +3158,87 @@ def _gradient_3d(floating[:, :, :] img, double[:, :] img_world2grid,
-                        # Compute coordinates of point dx on img's grid
-                        dx[p] = x[p] - h[p]
-                        q[0] = _apply_affine_3d_x0(dx[0], dx[1], dx[2], 1,
-                                                   img_world2grid)
-                        q[1] = _apply_affine_3d_x1(dx[0], dx[1], dx[2], 1,
-                                                   img_world2grid)
-                        q[2] = _apply_affine_3d_x2(dx[0], dx[1], dx[2], 1,
-                                                   img_world2grid)
-                        # Interpolate img at q
-                        in_flag = _interpolate_scalar_3d[floating](img, q[0],
-                            q[1], q[2], &out[k, i, j, p])
-                        if in_flag == 0:
-                            out[k, i, j, p] = 0
-                            inside[k, i, j] = 0
-                            continue
-                        tmp = out[k, i, j, p]
-                        # Compute coordinates of point dx on img's grid
-                        dx[p] = x[p] + h[p]
-                        q[0] = _apply_affine_3d_x0(dx[0], dx[1], dx[2], 1,
-                                                   img_world2grid)
-                        q[1] = _apply_affine_3d_x1(dx[0], dx[1], dx[2], 1,
-                                                   img_world2grid)
-                        q[2] = _apply_affine_3d_x2(dx[0], dx[1], dx[2], 1,
-                                                   img_world2grid)
-                        # Interpolate img at q
-                        in_flag = _interpolate_scalar_3d[floating](img, q[0],
-                            q[1], q[2], &out[k, i, j, p])
-                        if in_flag == 0:
-                            out[k, i, j, p] = 0
-                            inside[k, i, j] = 0
-                            continue
-                        out[k, i, j, p] = (out[k, i, j, p] - tmp) / img_spacing[p]
-                        dx[p] = x[p]
+        for k in prange(nslices, schedule='guided'):
+            _gradient_3d_fun(img, img_world2grid, img_spacing,
+                             out_grid2world, out, inside, k)
+
+    if have_openmp and num_threads is not None:
+        openmp.omp_set_num_threads(all_cores)
+
+
+cdef void _sparse_gradient_3d_fun(floating[:, :, :] img, double[:, :] img_world2grid,
+                        double[:] img_spacing, double[:, :] sample_points,
+                        floating[:, :] out, int[:] inside, int i) nogil:
+    cdef:
+        int in_flag, p
+        double tmp, dx[3], h[3], q[3]
+    h[0] = 0.5 * img_spacing[0]
+    h[1] = 0.5 * img_spacing[1]
+    h[2] = 0.5 * img_spacing[2]
+    inside[i] = 1
+    dx[0] = sample_points[i, 0]
+    dx[1] = sample_points[i, 1]
+    dx[2] = sample_points[i, 2]
+    for p in range(3):
+        # Compute coordinates of point dx on img's grid
+        dx[p] = sample_points[i, p] - h[p]
+        q[0] = _apply_affine_3d_x0(dx[0], dx[1], dx[2], 1,
```

Comment:
I would rather see `_apply_affine_3d_x0` as a function where the last digit is also a parameter...